plugin/evm/validators_cache_test.go

```diff
@@ -0,0 +1,59 @@
+package evm
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/flare-foundation/coreth/tests/mocks"
+	"github.com/flare-foundation/flare/ids"
+)
+
+func TestWithCacheSize(t *testing.T) {
+	wantCacheSize := uint(42)
+
+	cfg := CacheConfig{}
+	WithCacheSize(wantCacheSize)(&cfg)
+
+	assert.Equal(t, wantCacheSize, cfg.CacheSize)
+}
+
+type TestValidatorCache struct {
+	ByEpochFunc func(uint64) (map[ids.ShortID]uint64, error)
+}
+
+func (tv TestValidatorCache) ByEpoch(e uint64) (map[ids.ShortID]uint64, error) {
+	return tv.ByEpochFunc(e)
+}
+
+func TestValidatorsCache_ByEpoch(t *testing.T) {
+	var data = mocks.RetrieverMock(5, 2)
+
+	tvc := TestValidatorCache{
+		ByEpochFunc: func(e uint64) (map[ids.ShortID]uint64, error) {
+			return data[e], nil
+		},
+	}
+
+	valCache := NewValidatorsCache(tvc, WithCacheSize(32))
+	for k, v := range data {
+		valCache.cache.Add(k, v)
+	}
+
+	t.Run("nominal case", func(t *testing.T) {
+		for _, i := range []uint64{0, 1} {
+			expected, _ := tvc.ByEpoch(i)
+
+			result, err := valCache.ByEpoch(i)
+			require.NoError(t, err)
+			assert.Equal(t, expected, result)
+		}
+	})
+
+	t.Run("missing key case", func(t *testing.T) {
+		result, err := valCache.ByEpoch(uint64(7))
+		require.NoError(t, err)
+		assert.Len(t, result, 0)
+	})
+}
```

Comment:
That's closer but still not exactly what I had in mind 😬 What I meant was more something along the lines of this:

* Only the nominal case contains the mocked function that returns actual legit data. The error case instead has a mock which returns errors.
	* As we discussed, in each test case, the mock logic is specified for that test specifically only.
* We add a test case for the situation where the mock returns an error.
* The calls to `mocks.RetrieverMock` are now made using named arguments that make it clear what 5 and 2 are in the context of the test
* The mock asserts that it gets the expected values as received arguments
* Instead of getting the expected result from the mock, we define it at the top of the test to make it clear to the reader what to expect. It is a bit more cumbersome than having something generic, but I'm not fond of retrieving the expected value from the mock personally.
* `tvc` is renamed to `mock` to emphasize on its purpose
* `data` is renamed to `testValidators`
* Some more small nitpicks

```suggestion
func TestValidatorsCache_ByEpoch(t *testing.T) {

	t.Run("nominal case", func(t *testing.T) {
		numValidators := 5
		numEpochs := 2
		testValidators := mocks.RetrieverMock(numValidators, numEpochs)
		want := map[ids.ShortID]uint64{
			// Fill this with the data we expect to get.
		}

		var calls int
		mock := TestValidatorCache{
			ByEpochFunc: func(e uint64) (map[ids.ShortID]uint64, error) {
				assert.Equal(t, calls, e)

				calls++
				return data[e], nil
			},
		}

		valCache := NewValidatorsCache(mock, WithCacheSize(32))
		for k, v := range testValidators {
			valCache.cache.Add(k, v)
		}

		result, err := valCache.ByEpoch(1)
		require.NoError(t, err)
		assert.Equal(t, expected, result)
	})

	t.Run("handles missing key", func(t *testing.T) {
		// I'm not sure how this test works in your case actually? Shouldn't it panic since we would attempt to get `validators[7]` in the map which does not exist?
		// I don't have time to look deeply at the code we're testing unfortunately so we might need to chat about it when I am back
	})

	t.Run("handles failure to retrieve validator by epoch", func(t *testing.T) {
		mock := TestValidatorCache{
			ByEpochFunc: func(uint64) (map[ids.ShortID]uint64, error) {
				return nil, errors.New("dummy error")
			},
		}

		valCache := NewValidatorsCache(mock, WithCacheSize(32))
	
		_, err := valCache.ByEpoch(uint64(7))
		assert.Error(t, err)
	})
}
```

Reply:
I have updated the tests https://github.com/flare-foundation/coreth/pull/25/commits/42101d73f668fc367336f11d03ee6a9b1b1f07f9
Please let me know what do you think